src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,131 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+
+        private static int LastUsedContractId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.TryGet(key);
+            return (int)(BigInteger)item - 1;
+        }
+
+        private static void SetNextAvailableContractId(DataCache snapshot, int newId)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            item.Set(newId);
+        }
+        
+        private static int GetNextAvailableId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            int value = (int)(BigInteger)item;
+            item.Add(1);
+            return value;
+        }
+
+        public static async Task<(ContractState contractState, IReadOnlyList<(string key, string value)> storagePairs)> DownloadContractStateAsync(
+            string contractHash,
+            string rpcUri,
+            uint stateHeight)
+        {
+            if (!UInt160.TryParse(contractHash, out var _contractHash))
+            {
+                throw new ArgumentException($"Invalid contract hash: \"{contractHash}\"");    
+            }
+
+            if (!TransactionExecutor.TryParseRpcUri(rpcUri, out var uri))
+            {
+                throw new ArgumentException($"Invalid RpcUri value \"{rpcUri}\"");
+            }
+
+            using var rpcClient = new RpcClient(uri);
+            var stateAPI = new StateAPI(rpcClient);
+                
+            uint height = stateHeight;
+            if (height == 0)
+            {
+                var stateHeight_ = await stateAPI.GetStateHeightAsync();
+                if (stateHeight_.localRootIndex is null)
+                {
+                    throw new Exception("Null \"localRootIndex\" in state height response");
+                }
+                height = stateHeight_.localRootIndex.Value;
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, new byte[0]);
+            var contractState = await rpcClient.GetContractStateAsync(contractHash).ConfigureAwait(false);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(SnapshotCache snapshot, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
```

Comment:
since the value returned from `GetAndChange` was passed to `GetInteroperable` and then discarded, I don't think updates to localContract will be reflected in storage

Reply:
I take it back, it appears to work just fine.